transport/x/cherami/inbound.go

```diff
@@ -0,0 +1,204 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cherami
+
+import (
+	"context"
+	"log"
+	"time"
+
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/internal/errors"
+	"go.uber.org/yarpc/internal/sync"
+	"go.uber.org/yarpc/serialize"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/uber/cherami-client-go/client/cherami"
+)
+
+const (
+	transportName = "cherami"
+
+	defaultPrefetchCount = 10
+
+	defaultCheramiTimeoutInSec = 15
+)
+
+// InboundConfig defines the config in order to create a Inbound
+// if frontend is provided, we'll connect to the provided frontend
+// otherwise, hyperbahn will be used to connect to cherami
+// PrefetchCount controls the number of messages to buffer locally.
+// Inbounds which process messages very fast may want to specify larger value
+// for PrefetchCount for faster throughput.  On the flip side larger values for
+// PrefetchCount will result in more messages being buffered locally causing high memory foot print
+
+type InboundConfig struct {
+	Destination         string
+	ConsumerGroup       string
+	Frontend            string
+	Port                int
+	PrefetchCount       int
+	CheramiTimeoutInSec int
+}
+
+// Inbound is a inbound that uses cherami as the transport
+type Inbound struct {
+	config         InboundConfig
+	consumer       cherami.Consumer
+	router         transport.Router
+	tracer         opentracing.Tracer
+	cheramiFactory CheramiFactory
+
+	once sync.LifecycleOnce
+}
+
+// NewInbound builds a new cherami inbound
+func NewInbound(config InboundConfig) *Inbound {
+	if config.PrefetchCount == 0 {
+		config.PrefetchCount = defaultPrefetchCount
+	}
+	if config.CheramiTimeoutInSec == 0 {
+		config.CheramiTimeoutInSec = defaultCheramiTimeoutInSec
+	}
+	return &Inbound{
+		config:         config,
+		tracer:         opentracing.GlobalTracer(),
+		cheramiFactory: NewCheramiFactory(),
+	}
+}
+
+// Transports returns nil for now
+func (i *Inbound) Transports() []transport.Transport {
+	return nil
+}
+
+// SetRouter configures a router to handle incoming requests.
+// This satisfies the transport.Inbound interface, and would be called
+// by a dispatcher when it starts.
+func (i *Inbound) SetRouter(router transport.Router) {
+	i.router = router
+}
+
+// IsRunning returns whether the inbound is still processing requests.
+func (i *Inbound) IsRunning() bool {
+	return i.once.IsRunning()
+}
+
+// Start starts the inbound, reading and handling messages from cherami
+func (i *Inbound) Start() error {
+	return i.once.Start(i.start)
+}
+
+func (i *Inbound) start() error {
+	if i.router == nil {
+		return errors.ErrNoRouter
+	}
+
+	var client cherami.Client
+	var err error
+	if len(i.config.Frontend) > 0 {
+		client, err = i.cheramiFactory.GetClientWithFrontEnd(i.config.Frontend, i.config.Port)
```

Comment:
Instead of creating a client directly from within the transport, we should *inject* the Cherami client into the transport:

```golang
client := cherami.NewClient() // where this is managed by rpcinit

cheramiTransport := cherami.NewTransport(cherami.Config{Client: client})

inbounds := []Inbounds{
  cheramiTransport.NewInbound(cherami.InboundConfig{...}},
  cheramiTransport.NewInbound(cherami.InboundConfig{...}},
  ...
}
outbounds := []Outbounds{
  cheramiTransport.NewOutbound{cherami.OutboundConfig{...}},
  ...
}
```
This will make this transport have very few reasons to change, and we can instead manage configuring Cherami clients in rpcinit:

1. An internal lib, like rpcinit, configures the Cherami client in an opinionated way
2. The client is injected into the Cherami transport

This allows the `CheramiFactory` to be moved to rpcinit, removing that opinionated config from here.